File: leetcode/problems/language-specific/go/1721.swapping-nodes-in-a-linked-list.go
/*
 * @lc app=leetcode id=1721 lang=golang
 *
 * [1721] Swapping Nodes in a Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapNodes(head *ListNode, k int) *ListNode {
	// 1. Find the k-th node from the front which is left
	// 2. Make left node as the current node, right node from the front,
	//    when the current node reach end, right node is just the k-th last element
	// 3. Swap their values

	left, right, curCount := head, head, 0
	for left != nil {
		curCount++
		if curCount == k {
			break
		}

		left = left.Next
	}

	cur := left
	for cur.Next != nil {
		cur = cur.Next
		right = right.Next
	}

	left.Val, right.Val = right.Val, left.Val
	return head
}

// @lc code=end

File: leetcode/problems/problem-specific/1721.swapping-nodes-in-a-linked-list/1721.swapping-nodes-in-a-linked-list.py
#
# @lc app=leetcode id=1721 lang=python3
#
# [1721] Swapping Nodes in a Linked List
#

# @lc code=start

# * Two Pointers Solution | O(n) Time | O(1) Space
# * n -> The number of nodes in the linked list


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        cur = head
        for _ in range(k - 1):
            cur = cur.next

        left, right = cur, head
        while cur.next:
            cur = cur.next
            right = right.next

        left.val, right.val = right.val, left.val
        return head


# @lc code=end
